rfcs/0110-inherit-as-list.md

```diff
@@ -0,0 +1,135 @@
+are more verbose and complex than `with`, they are rarely used.
+
+The goal of this RFC is to provide a similarly-terse alternative which avoids
+these drawbacks.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed syntax is:
+
+```
+attrs.[ a b c ]
+```
+
+This expression is syntactic sugar for:
+
+```nix
+[ attrs.a attrs.b attrs.c ]
+```
+
+As the token `.` immediately preceding `[` is currently a syntax error,
+a Nix interpreter which supports this new language feature will be compatible
+with existing Nix code.
+
+This RFC is implemented here: https://github.com/nixos/nix/pull/5402
+
+For MVP functionality, only minor additions to `src/libexpr/parser.y` are
+needed. If accepted, the changes to the Nix interpreter can be backported
+to current releases if desired. Relevant language documentation and
+third-party parsers/linters would also need to be updated.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+This would be useful for many languages and frameworks in Nixpkgs which
+extract packages from a package set argument.
+
+For example, `python3.withPackages (ps: ps.[ ... ])` will serve as a
+more fine-grained alternative to `python3.withPackages (ps: with ps; [ ... ])`.
+This would apply similarly to `vim.withPlugins`, `lua.withPackages`, etc.
+
+Certain list-typed `meta` fields could also make use of this feature, e.g.:
+```
+meta.licenses = lib.licenses.[ bsd3 mit ];
+meta.maintainers = lib.maintainers.[ johndoe janedoe ];
+```
+
+Note that only simple textual attrnames are allowed in the square brackets.
+For example, `pkgs.[ (openssl.overrideAttrs { patches = [ ... ]; }) ]`
+is currently a syntax error, as is `pkgs.[ "${some_expression}" ]`,
+`a.[ b.[ c d ] ]`, and `a.[ [ b c ] [ d e ] ]`.
+Future RFCs may add additional support for useful idioms such as
+`pkgs.[ python310 python310Packages.pytorch ]` on a case-by-case basis,
+but that is not planned for this RFC.
+
+Other forms of syntax which were considered but are not proposed
+in this RFC include:
+* `[ inherit (attrs) a b c; ]`
+* `[ inherit a b c; ]`
+* `inherit (attrs) [ a b c ]`
+* `pkgs.{ python = python310; openssl = openssl_3; }`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This will add complexity to the Nix grammar and any third-party tools which
+  operate on Nix expressions.
+* Expressions reliant on the new syntax will be incompatible with
+  Nix versions prior to the introduction of this feature.
+
+# Alternatives
+[alternatives]: #alternatives
+
+* Skillful use of `with` to avoid its drawbacks
+* Fine-grained attribute selection via existing (more verbose) language
+  features, such as `builtins.attrValues { inherit (attrs) a b c; }`
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+How would this feature be adopted, if accepted?
+
+# Future work
+[future]: #future-work
+
+Determine best practices regarding when this language construct should be used
```

Comment:
Should we say something about adoption in nixpkgs? Probably disallow its usage for the foreseeable future until a majority is using a version of Nix with support for inherit-as-list.

Reply:
Should the feature be back ported to (for instance) 2.4.1, so that adoption will cost less time?

Reply:
Should these kinds of notes be part of the future work section of the RFC?

Reply:
I am not sure whether a new feature like this qualifies for backporting, but the currently proposed implementation is very simple, so it would be easy to backport to any versions we desire.

Reply:
Backwards compatibility is very important for nixpkgs, and there are still a lot of users on older Nix versions, so I'd definitely not propose adoption in nixpkgs as part of this RFC. I'd leave that question to a future RFC.

Reply:
When this is implemented in Nix, it's easy for this feature to be used in Nixpkgs and slip through the review process.

It might be useful to describe a systematic way in this RFC to avoid this feature in nixpkgs until the other RFC is picked up.

Not doing anything with this in this RFC seems like it will cause problems.

Reply:
Hmm, good point. It seems like the experimental-features flag would be good for this, but I think there was some opposition to gating this syntax behind an experimental feature. Maybe we should revisit that option?

Reply:
What people want to do in their own repositories is something they may decide for themselves. The experimental flag will cause problems when sharing such repositories. Anyone contributing to the repo needs to enable the flag: the flag will then also be enabled when contributing to nixpkgs. It's messy and confusing especially for new people. The flag is forces upon you and eventually the older community members will converge to having it enabled, just like what happened to flakes. Newcomers will get surprised.

Personally I think nixpkgs lacks a linter. There it is possible to already support the new feature, but show a linting error with proper description why it is not allowed. I'm guessing this requires a separate RFC 😅

With a linter it will also be possible to guard against confusing usage of `with`.

When the new feature is adopted in nixpkgs, a linter can suggest the new feature as an alternative to usage of `with`.

Reply:
@bobvanderlinden 
> With a linter it will also be possible to guard against confusing usage of with.

Do you come up with the pattern of "confusing with"? I could also help with [my language server](https://github.com/oxalica/nil). `let foo = ...; in with ...; [ foo ]` would make too many false-positives, because that's the only way to access variables outside `with`.